structopt-derive/src/attrs.rs

```diff
@@ -341,155 +356,207 @@ impl Attrs {
                 } else {
                     arg
                 };
-                res.push_str_method(name, &new_arg);
+                let name = Sp::call_site(name.to_string());
+                let new_arg = Sp::call_site(new_arg.to_string());
+                res.push_str_method(name, new_arg);
             });
         res.push_doc_comment(attrs, "about");
         res.push_attrs(attrs);
         if res.has_custom_parser {
-            panic!("parse attribute is only allowed on fields");
+            span_error!(
+                res.parser.span(),
+                "parse attribute is only allowed on fields"
+            );
         }
-        match res.kind {
-            Kind::Subcommand(_) => panic!("subcommand is only allowed on fields"),
-            Kind::FlattenStruct => panic!("flatten is only allowed on fields"),
-            Kind::Skip => panic!("skip is only allowed on fields"),
+        match &*res.kind {
+            Kind::Subcommand(_) => {
+                span_error!(res.kind.span(), "subcommand is only allowed on fields")
+            }
+            Kind::FlattenStruct => {
+                span_error!(res.kind.span(), "flatten is only allowed on fields")
+            }
+            Kind::Skip => span_error!(res.kind.span(), "skip is only allowed on fields"),
             Kind::Arg(_) => res,
         }
     }
-    fn ty_from_field(ty: &syn::Type) -> Ty {
+    fn ty_from_field(ty: &syn::Type) -> Sp<Ty> {
+        let t = |kind| Sp::new(kind, ty.span());
         if let Path(TypePath {
             path: syn::Path { ref segments, .. },
             ..
         }) = *ty
         {
             match segments.iter().last().unwrap().ident.to_string().as_str() {
-                "bool" => Ty::Bool,
+                "bool" => t(Ty::Bool),
                 "Option" => match sub_type(ty).map(Attrs::ty_from_field) {
-                    Some(Ty::Option) => Ty::OptionOption,
-                    Some(Ty::Vec) => Ty::OptionVec,
-                    _ => Ty::Option,
+                    Some(ty) => match *ty {
+                        Ty::Option => t(Ty::OptionOption),
+                        Ty::Vec => t(Ty::OptionVec),
+                        _ => t(Ty::Option),
+                    },
```

Comment:
minor: I think you can map to flatten this match

Reply:
Done